sysutils/apcupsd/src/opnsense/www/js/widgets/Apcupsd.js

```diff
@@ -0,0 +1,331 @@
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import BaseTableWidget from "./BaseTableWidget.js";
+
+/**
+ * @typedef ApcUpsdStatusValue
+ *
+ * @property {string|number} norm - normalized value
+ * @property {string|number} value - raw value
+ */
+
+/**
+ * @typedef {Object.<string, ApcUpsdStatusValue>} ApcUpsdStatus
+ */
+
+export default class ApcUpsd extends BaseTableWidget {
+    constructor() {
+        super();
+    }
+
+    getMarkup() {
+        let $container = $('<div></div>');
+        let $apcupsdTable = this.createTable('apcupsd-table', {
+            headerPosition: 'left'
+        });
+        $container.append($apcupsdTable);
+        return $container;
+    }
+
+    async onWidgetTick() {
+        const data = await this.ajaxCall('/api/apcupsd/service/getUpsStatus');
+
+        if (data.error) {
+            this.displayError(data.error);
+            return;
+        }
+
+        let rows = this.makeRows(data.status);
+        super.updateTable('apcupsd-table', rows);
+    }
+
+    /**
+     * Displays an error text row
+     *
+     * @param {string} text
+     */
+    displayError(text) {
+        // A row with `flextable-row` class will be auto removed on next tick by
+        // `this.updateTable(...)`, if the error is gone
+        const $row = $('<div class="flextable-row text-danger" style="justify-content: center"></div>').text(text);
+        // Replace all the content in the table
+        $(`#apcupsd-table`).html($row);
+    }
+
+    /**
+     * @param {ApcUpsdStatus} status
+     * @returns {Array<Array>>}
+     */
+    makeRows(status) {
+        const rows = [];
+
+        if (status.MODEL) {
+            rows.push(this.makeTextRow("status_model", status.MODEL));
+        }
+
+        if (status.STATUS) {
+            rows.push(
+                // Note: text value is not translated because it comes directly from apcupsd output
+                this.makeColoredTextRow(
+                    "status_status",
+                    status.STATUS.value,
+                    /ONLINE/, // success
+                    /ONBATT|LOWBATT|REPLACEBATT|NOBATT|SLAVEDOWN|COMMLOST/ // error
+                )
+            );
+        }
+
+        if (status.SELFTEST) {
+            const raw_value = status.SELFTEST.value;
+
+            // self test values
+            const value_keys = {
+                OK: 'status_selftest_ok',
+                BT: 'status_selftest_battery',
+                NG: 'status_selftest_overload',
+                NO: 'status_selftest_no_results',
```

Comment:
These do not match with translation keys

Reply:
```suggestion
                BT: 'status_selftest_bt',
                NG: 'status_selftest_ng',
                NO: 'status_selftest_no',
```

Reply:
thanks, nice catch